### PR TITLE
pythonPackages.transformers: init at 2.2.1

### PR DESCRIPTION
**Pull request by pashashocky**
Adding the [transformers](https://github.com/huggingface/transformers) library for DeepLearning along side the dependencies to build it.

###### Motivation for this change
Wanted to build the transformers library for ML work.

This is my first PR into NixOS, so please let me know if I am doing something wrong and I will try to rectify this in the future, but I struggled quite a bit through this - so happy that it builds. This builds on my system and made sure this builds with Sandboxing. 

Wasn't sure if I had to create multiple PR's for each separate package that transformers depends on, but it's all in here. 

Tests for transformers exist, but i don't think they will work in sandboxing because they require downloading packages from s3, and also require to have tensorflow or pytorch installed... I will try to get them to work one more time, but failed on first try.

###### Things done
- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

###### Notify maintainers

cc @jonringer @teh @FRidh - Not sure who I am supposed to be tagging for PR's that don't have any maintainers.

### CODE REVIEW

File: maintainers/maintainer-list.nix [hunk 1]
```diff
@@ -5187,6 +5187,12 @@
     githubId = 20792;
     name = "Sebastian Galkin";
   };
+  pashashocky = {
+    email = "pashashocky@gmail.com";
+    github = "pashashocky";
+    githubId = 673857;
+    name = "Pash Shocky";
+  };
```
Comment (jonringer) [inline maintainers/maintainer-list.nix]: please make this a separate commit
```
maintainers: add pashashocky
```

File: pkgs/development/libraries/sentencepiece/default.nix [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+, stdenv
+, lib
+, fetchgit
+, cmake
+, gperftools
+}:
+
+stdenv.mkDerivation rec {
+  pname = "sentencepiece";
+  version = "0.1.84";
```
Comment (jonringer) [inline pkgs/development/libraries/sentencepiece/default.nix]: this should be a separate commit
```
sentencepiece: init at 0.1.84
```

File: pkgs/development/libraries/sentencepiece/default.nix [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+
+  src = fetchgit {
```
Comment (jonringer) [inline pkgs/development/libraries/sentencepiece/default.nix]: please use fetchFromGitHub
```suggestion
  src = fetchFromGitHub {
    owner = "google";
    repo = pname;
    rev = "v${version";
    sha = ....
```

File: pkgs/development/python-modules/sacremoses/default.nix [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+, fetchgit
+, click
+, six
+, tqdm
+, joblib
+}:
+
+buildPythonPackage rec {
+  pname = "sacremoses";
+  version = "0.0.35";
```
Comment (jonringer) [inline pkgs/development/python-modules/sacremoses/default.nix]: this should be a separate commit

File: pkgs/development/python-modules/sacremoses/default.nix [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+
+  src = fetchgit {
```
Comment (jonringer) [inline pkgs/development/python-modules/sacremoses/default.nix]: please use fetchFromGitHub

File: pkgs/development/python-modules/sacremoses/default.nix [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+    url = "https://github.com/alvations/sacremoses";
+    rev = "refs/tags/${version}";
+    sha256 = "1gzr56w8yx82mn08wax5m0xyg15ym4ri5l80gmagp8r53443j770";
+  };
+
+  doCheck = false;
```
Comment (jonringer) [inline pkgs/development/python-modules/sacremoses/default.nix]: please say why tests aren't enabled

File: pkgs/development/python-modules/sentencepiece/default.nix [hunk 7]
```diff
@@ -0,0 +1,17 @@
+{ buildPythonPackage
+, stdenv
+, fetchgit
```
Comment (jonringer) [inline pkgs/development/python-modules/sentencepiece/default.nix]: unneeded
```suggestion
```

File: pkgs/development/python-modules/sentencepiece/default.nix [hunk 8]
```diff
@@ -0,0 +1,17 @@
...
+, sentencepiece
+, pkgs
```
Comment (jonringer) [inline pkgs/development/python-modules/sentencepiece/default.nix]: import on a per-packagke basis
```suggestion
, pkgconfig
```

File: pkgs/development/python-modules/transformers/default.nix [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+, tqdm
+, regex
+, numpy
+}:
+
+buildPythonPackage rec {
+  pname = "transformers";
+  version = "2.2.1";
+
+  src = fetchgit {
```
Comment (jonringer) [inline pkgs/development/python-modules/transformers/default.nix]: please use fetchFromGitHub

File: pkgs/top-level/python-packages.nix [hunk 10]
```diff
@@ -6593,6 +6593,14 @@ in {
 
   pony = callPackage ../development/python-modules/pony { };
 
+  sentencepiece = callPackage ../development/python-modules/sentencepiece {
+    sentencepiece = pkgs.sentencepiece;
```
Comment (jonringer) [inline pkgs/top-level/python-packages.nix]:
```suggestion
    inherit (pkgs) sentencepiece pkgconfig;
```

File: pkgs/top-level/python-packages.nix [hunk 11]
```diff
@@ -6593,6 +6593,14 @@ in {
...
+  };
+
+  transformers = callPackage ../development/python-modules/transformers { };
+
+  sacremoses = callPackage ../development/python-modules/sacremoses { };
```
Comment (jonringer) [inline pkgs/top-level/python-packages.nix]: please attempt to order these

**Comment by pashashocky** [comment]
@jonringer - thank you for your comments, fixed and rebased.

File: pkgs/development/python-modules/sentencepiece/default.nix [hunk 12]
```diff
@@ -0,0 +1,16 @@
...
+
+buildPythonPackage rec {
+  pname = "sentencepiece";
+  inherit (sentencepiece) version src meta;
+
+  postPatch = ''
+    cd python
+  '';
+  buildInputs = [ sentencepiece ];
+  nativeBuildInputs = [ pkgconfig ];
```
Comment (jonringer) [inline pkgs/development/python-modules/sentencepiece/default.nix]: usually inputs come after src, and then ordered by more host-specific to target-specific inputs

here's how I would write this:
```
  nativeBuildInputs = [ pkgconfig ];
  buildInputs = [ sentencepiece ];

  sourceRoot = "$sourceRoot/python"
```

File: pkgs/development/python-modules/sentencepiece/default.nix [hunk 13]
```diff
@@ -0,0 +1,16 @@
+{ buildPythonPackage
+, stdenv
+, sentencepiece
+, pkgconfig
+}:
+
+buildPythonPackage rec {
+  pname = "sentencepiece";
+  inherit (sentencepiece) version src meta;
```
Comment (jonringer) [inline pkgs/development/python-modules/sentencepiece/default.nix]: I would probably break this out into a separate commit as well:
```
pythonPackages.sentencepiece: init at 0.1.84
```

File: pkgs/development/python-modules/sentencepiece/default.nix (hunk 12 above)
Comment (pashashocky) [reply pkgs/development/python-modules/sentencepiece/default.nix]: It doesn't seem to recognize "${sourceRoot}/python":
```
error: while evaluating the attribute 'buildInputs' of the derivation 'nix-shell' at /home/pash/nixpkgs/pkgs/build-support/mkshell/default.nix:28:3:
while evaluating the attribute 'passAsFile' of the derivation 'python3-3.7.5-env' at /home/pash/nixpkgs/pkgs/build-support/trivial-builders.nix:7:14:
while evaluating 'requiredPythonModules' at /home/pash/nixpkgs/pkgs/top-level/python-packages.nix:64:27, called from /home/pash/nixpkgs/pkgs/development/interpreters/python/wrapper.nix:15:13:
while evaluating 'unique' at /home/pash/nixpkgs/lib/lists.nix:643:12, called from /home/pash/nixpkgs/pkgs/top-level/python-packages.nix:66:6:
while evaluating 'requiredPythonModules' at /home/pash/nixpkgs/pkgs/top-level/python-packages.nix:64:27, called from /home/pash/nixpkgs/pkgs/top-level/python-packages.nix:82:33:
while evaluating 'unique' at /home/pash/nixpkgs/lib/lists.nix:643:12, called from /home/pash/nixpkgs/pkgs/top-level/python-packages.nix:66:6:
while evaluating 'unique' at /home/pash/nixpkgs/lib/lists.nix:643:12, called from /home/pash/nixpkgs/lib/lists.nix:649:17:
while evaluating anonymous function at /home/pash/nixpkgs/lib/lists.nix:152:16, called from undefined position:
while evaluating the attribute 'out.outPath' at /home/pash/nixpkgs/lib/customisation.nix:151:13:
while evaluating the attribute 'sourceRoot' of the derivation 'python3.7-sentencepiece-0.1.84' at /home/pash/nixpkgs/pkgs/development/interpreters/python/mk-python-derivation.nix:102:3:
infinite recursion encountered, at /home/pash/nixpkgs/pkgs/development/python-modules/sentencepiece/default.nix:14:19
```

looking into it

Comment (jonringer) [reply pkgs/development/python-modules/sentencepiece/default.nix]: you don't want to do `${sourceRoot}/python"; you just want `$sourceRoot/python`;

Comment (jonringer) [reply pkgs/development/python-modules/sentencepiece/default.nix]: `${sourceRoot}` needs to be evaluated by nix, but we are defining sourceRoot, so it's referencing itself, thus the infinite recursion. If we just do the bash interpolation, then it just gets evaluated once

Comment (pashashocky) [reply pkgs/development/python-modules/sentencepiece/default.nix]: @jonringer yeah, i tried that too and it can't find the directory in that case: 

```
these derivations will be built:
  /nix/store/4k9rln87gj7sz91vh1hlgsqk4ard45gf-python3.7-sentencepiece-0.1.84.drv
  /nix/store/7q70rczzy68mfld2xh8cw4r80qci9s3p-python3.7-transformers-2.2.1.drv
  /nix/store/w3j539ak3h54sbfvp7w2r4gplag160h7-python3-3.7.5-env.drv
building '/nix/store/4k9rln87gj7sz91vh1hlgsqk4ard45gf-python3.7-sentencepiece-0.1.84.drv'...
Sourcing python-catch-conflicts-hook.sh
Sourcing python-remove-bin-bytecode-hook.sh
Sourcing setuptools-build-hook
Using setuptoolsBuildPhase
Using setuptoolsShellHook
Sourcing pip-install-hook
Using pipInstallPhase
Sourcing python-imports-check-hook.sh
Using pythonImportsCheckPhase
Sourcing setuptools-check-hook
Using setuptoolsCheckPhase
unpacking sources
unpacking source archive /nix/store/y7a0kjka6gdlfdrqq9i32g78lvsc2287-source
source root is $sourceRoot/python
chmod: cannot access '$sourceRoot/python': No such file or directory
builder for '/nix/store/4k9rln87gj7sz91vh1hlgsqk4ard45gf-python3.7-sentencepiece-0.1.84.drv' failed with exit code 1
cannot build derivation '/nix/store/w3j539ak3h54sbfvp7w2r4gplag160h7-python3-3.7.5-env.drv': 1 dependencies couldn't be built
error: build of '/nix/store/w3j539ak3h54sbfvp7w2r4gplag160h7-python3-3.7.5-env.drv' failed
```

Comment (pashashocky) [reply pkgs/development/python-modules/sentencepiece/default.nix]: this works though: 

```
  nativeBuildInputs = [ pkgconfig ];
  buildInputs = [ sentencepiece ];

  postUnpack = "sourceRoot=$sourceRoot/python";

```

Comment (pashashocky) [reply pkgs/development/python-modules/sentencepiece/default.nix]: this also works: 

```
  nativeBuildInputs = [ pkgconfig ];
  buildInputs = [ sentencepiece ];

  sourceRoot = "source/python";

```

Comment (pashashocky) [reply pkgs/development/python-modules/sentencepiece/default.nix]: I went with the latter... rebased and fixed.

Comment (jonringer) [reply pkgs/development/python-modules/sentencepiece/default.nix]: yea, I usually use the latter as well. I thought you would be able to use what I recommended, guess I should have tested first.

Comment (jonringer) [reply pkgs/development/python-modules/sentencepiece/default.nix]: actually, that makes sense, when it evaluates the derivation, that's when it will set the sourceRoot value, which doesn't exist, so if you do the bash substitution like i recommend it would just be empty so `sourceRoot="$sourceRoot/python"` would be just "/python", which isn't right. Sorry for leading you astray :(

File: pkgs/development/python-modules/sacremoses/default.nix [hunk 14]
```diff
@@ -0,0 +1,29 @@
...
+
+  src = fetchFromGitHub {
+    owner = "alvations";
+    repo = pname;
+    rev = "${version}";
+    sha256 = "1gzr56w8yx82mn08wax5m0xyg15ym4ri5l80gmagp8r53443j770";
+  };
+
+  propagatedBuildInputs = [ click six tqdm joblib ];
+
```
Comment (jonringer) [inline pkgs/development/python-modules/sacremoses/default.nix]: do you mind adding tests? their important to ensure that you aren't broken by dependency bumps

```

  checkInputs = [ pytest ];
  # ignore tests which call to remote host
  checkPhase = ''
    pytest -k 'not truecase'
  '';
```

File: pkgs/development/python-modules/transformers/default.nix [hunk 15]
```diff
@@ -0,0 +1,36 @@
...
+  src = fetchFromGitHub {
+    owner = "huggingface";
+    repo = pname;
+    rev = "v${version}";
+    sha256 = "1p8p3lhhiyk1xl9gpgq4vbchyz57v3w7hhvsj1r90zs3cckindl8";
+  };
+
+  # Tests disabled because requires downloading additional files from s3
+  # as well as having tensorflow and/or pytorch installed to perform them.
+  doCheck = false;
```
Comment (jonringer) [inline pkgs/development/python-modules/transformers/default.nix]: please enable

```
  checkInputs = [ pytest ];
  # pretrained tries to download from s3
  checkPhase = ''
    cd transformers # avoid importing local files
    HOME=$TMPDIR pytest -k 'not pretrained_tokenizers'
  '';
```

Comment (pashashocky) [reply pkgs/development/python-modules/transformers/default.nix]: yup, done - thank you. 

I seem to be missing a trick here, how did you quickly find the substring for which tests to exclude? For both transformers and sacremoses? Thanks!

Comment (pashashocky) [reply pkgs/development/python-modules/transformers/default.nix]: I also had a few more questions about general nix development patterns, like how to best keep your system on one release and have another repo for development, but be able to use these derivations straight after commiting? 

And a few others... is there a recommended place to ask them or we can talk somewhere? Thank you!

Comment (jonringer) [reply pkgs/development/python-modules/transformers/default.nix]: the -k arguments will just filter based on a substring of any test. I just ran the entire test suite, saw which failed, and determined whether it was an issue due to the sandbox environment or an issue with the package. In this case, that one test failure just because of the sandbox environment which is fine.

**Comment by jonringer** [comment]
@GrahamcOfBorg build python27Packages.sacremoses python27Packages.sentencepiece python27Packages.transformers python37Packages.sacremoses python37Packages.sentencepiece python37Packages.transformers python38Packages.sacremoses python38Packages.sentencepiece python38Packages.transformers sentencepiece

**Comment by jonringer** [comment]
do you mind updating the platforms for the packages? I'm trying to not have broken arch or darwin packages get added.

It's fine if this package just doesn't support those platforms

**Comment by pashashocky** [comment]
@jonringer - done, *hopefully correctly*, once again, appreciate your help!

**Review by jonringer — APPROVED** [review]
diff LGTM
commits LGTM

```
[14 built, 4 copied (8.8 MiB), 1.4 MiB DL]
https://github.com/NixOS/nixpkgs/pull/75106
10 package were built:
python27Packages.sacremoses python27Packages.sentencepiece python27Packages.transformers python37Packages.sacremoses python37Packages.sentencepiece python37Packages.transformers python38Packages.sacremoses python38Packages.sentencepiece python38Packages.transformers sentencepiece
```

**Comment by jonringer** [comment]
@GrahamcOfBorg build python27Packages.sacremoses python27Packages.sentencepiece python27Packages.transformers python37Packages.sacremoses python37Packages.sentencepiece python37Packages.transformers python38Packages.sacremoses python38Packages.sentencepiece python38Packages.transformers sentencepiece

**Comment by jonringer** [comment]
been busy with real life + oni, sorry for not getting back to this faster

File: pkgs/development/libraries/sentencepiece/default.nix [hunk 16]
```diff
@@ -0,0 +1,31 @@
...
+
+  enableParallelBuilding = true;
+
+  nativeBuildInputs = [ cmake gperftools ];
+
+  meta = with stdenv.lib; {
+    homepage = https://github.com/google/sentencepiece;
+    description = "Unsupervised text tokenizer for Neural Network-based text generation";
+    license = licenses.asl20;
+    platforms = platforms.linux;
```
Comment (jonringer) [inline pkgs/development/libraries/sentencepiece/default.nix]:
```suggestion
    platforms = [ "x86_64-linux" ];
```

the python package fails to build on all interpreter versions
```
cannot build derivation '/nix/store/8md1vqnkv98mm67q04xwwbgcr38aq6lp-python2.7-transformers-2.2.1.drv': 1 dependencies couldn't be built
/nix/store/m7lk7g5ccvg7x4h0jwigcb4lpxwzgk8a-setuptools-check-hook/nix-support/setup-hook: line 4:    75 Segmentation fault      (core dumped) /nix/store/9894fxjpg8v99j14kip6b13rdfgf4m5k-python3-3.7.5/bin/python3.7 nix_run_setup test
/nix/store/i3zis1yv4srhb62asn0442xjsf9ry0hl-setuptools-check-hook/nix-support/setup-hook: line 4:    75 Segmentation fault      (core dumped) /nix/store/hywqavwx22kk56lvsl06ihhswfr89dcp-python3-3.8.0/bin/python3.8 nix_run_setup test
builder for '/nix/store/q4m51ghk3y3n2ih0v2y5ls2y3xj19vs1-python3.7-sentencepiece-0.1.84.drv' failed with exit code 139
cannot build derivation '/nix/store/s50pg0qqxg8sqsrjdxvhkkpm8l0x5m4c-python3.7-transformers-2.2.1.drv': 1 dependencies couldn't be built
builder for '/nix/store/7bx6246na5r2s1jlwsy69siv7c1fw41d-python3.8-sentencepiece-0.1.84.drv' failed with exit code 139
cannot build derivation '/nix/store/xdhqracb4ib4dzqc9p7db50mp4xphv9y-python3.8-transformers-2.2.1.drv': 1 dependencies couldn't be built
error: build of '/nix/store/1iv6mnw2zlypi6fvnjzfq72fjv8gr685-python2.7-sentencepiece-0.1.84.drv', '/nix/store/7bx6246na5r2s1jlwsy69siv7c1fw41d-python3.8-sentencepiece-0.1.84.drv', '/nix/store/8md1vqnkv98mm67q04xwwbgcr38aq6lp-python2.7-transformers-2.2.1.drv', '/nix/store/q4m51ghk3y3n2ih0v2y5ls2y3xj19vs1-python3.7-sentencepiece-0.1.84.drv', '/nix/store/s50pg0qqxg8sqsrjdxvhkkpm8l0x5m4c-python3.7-transformers-2.2.1.drv', '/nix/store/xdhqracb4ib4dzqc9p7db50mp4xphv9y-python3.8-transformers-2.2.1.drv' failed
```

**Review by jonringer — CHANGES_REQUESTED** [review]

File: pkgs/development/libraries/sentencepiece/default.nix (hunk 16 above)
Comment (pashashocky) [reply pkgs/development/libraries/sentencepiece/default.nix]: done

Comment (pashashocky) [reply pkgs/development/libraries/sentencepiece/default.nix]: erm - think need to do for all parts as well - one sec

Comment (pashashocky) [reply pkgs/development/libraries/sentencepiece/default.nix]: @jonringer am i missing something with the checks? aarch64-linux seems to be failing for some reason...

**Comment by jonringer** [comment]
@GrahamcOfBorg build python27Packages.sacremoses python27Packages.sentencepiece python27Packages.transformers python37Packages.sacremoses python37Packages.sentencepiece python37Packages.transformers python38Packages.sacremoses python38Packages.sentencepiece python38Packages.transformers sentencepiece

**Review by jonringer — APPROVED** [review]

**Comment by jonringer** [comment]
platformo issues resolved

**Comment by pashashocky** [comment]
@jonringer Thank you for your help! I learned a lot!
